tests/apollo/test_skvbc_chaotic_startup.py

```diff
@@ -237,24 +218,114 @@ async def test_f_staggered_replicas_requesting_vc(self, bft_network, skvbc, nurs
+            print("STATUS: Early replicas that have initiated View Change catch up on state.")
+
+        # We stop the current primary and let the system to install a new view, while assuming the identity of
+        # the next primary
+        bft_network.stop_replica(initial_primary)
+
+        view = await bft_network.wait_for_view(
+            replica_id=expected_next_primary,
+            expected=lambda v: v == expected_next_view,
+            err_msg="Make sure we are in the next view "
+        )
+
+        self.assertTrue(expected_next_view == view)
+
+        await trio.sleep(5)  # TODO: remove when bft_network.wait_for_view also waits for system liveness.
+
+        await self._verify_replicas_are_in_view(view, bft_network.all_replicas(without={initial_primary}), bft_network)
+        await self._verify_the_system_is_able_to_make_progress(bft_network, skvbc, expected_next_primary, bft_network.all_replicas(without={initial_primary}))
+
+    async def _verify_replicas_are_in_view(self, view, replicas, bft_network):
+        for r in replicas:
             active_view_of_replica = await self._get_gauge(r, bft_network, 'currentActiveView')
             view_of_replica = await self._get_gauge(r, bft_network, 'view')
             self.assertTrue(active_view_of_replica == view_of_replica)  # verify Replica is not requesting View Change
             self.assertTrue(active_view_of_replica == view)  # verify Replica is in the current view
 
-        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=expected_next_primary)
+    async def _verify_the_system_is_able_to_make_progress(self, bft_network, skvbc, replica_to_read_from, initial_nodes, checkpoint_num=1, verify_checkpoint_persistency=False, assert_state_transfer_not_started=False):
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=replica_to_read_from)
         await skvbc.fill_and_wait_for_checkpoint(
-            initial_nodes=bft_network.all_replicas(without={replica_to_stop}),
-            checkpoint_num=1,
-            verify_checkpoint_persistency=False,
-            assert_state_transfer_not_started=False
+            initial_nodes=initial_nodes,
+            checkpoint_num=checkpoint_num,
+            verify_checkpoint_persistency=verify_checkpoint_persistency,
+            assert_state_transfer_not_started=assert_state_transfer_not_started
         )
-        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=expected_next_primary)
-
-        # Verify that the system is making progress after the View Change.
+        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=replica_to_read_from)
+        # Verify the system is able to make progress
         self.assertTrue(checkpoint_after > checkpoint_before)
 
+    async def _wait_for_less_than_f_plus_one_replicas_to_initiate_viewchange(self, replicas, bft_network):
```

Comment:
why not make this more generic:
- call the method `_wait_for_replicas_to_initiate_view_change()`
- remove the `self.assertTrue(num_of_replicas <= bft_network.config.f)`

Reply:
This method would work deterministically only if the view change is run by less than f+1 replicas. Otherwise, the replicas may complete the view change and then the halting condition will never occur.